test/v200/utils/common/test_utils.go

```diff
@@ -0,0 +1,283 @@
+
+// GetRandomString returns a random string which is n characters long.
+// If lower is set to true a lower case string is returned.
+func GetRandomString(n int, lower bool) string {
+	b := make([]byte, n)
+	for i := range b {
+		b[i] = schemaBytes[rand.Intn(len(schemaBytes)-1)]
+	}
+	randomString := string(b)
+	if lower {
+		randomString = strings.ToLower(randomString)
+	}
+	return randomString
+}
+
+var GroupKinds = [...]schema.CommandGroupKind{schema.BuildCommandGroupKind, schema.RunCommandGroupKind, schema.TestCommandGroupKind, schema.DebugCommandGroupKind}
+
+// GetRandomGroupKind return random group kind. One of "build", "run", "test" or "debug"
+func GetRandomGroupKind() schema.CommandGroupKind {
+	return GroupKinds[GetRandomNumber(1, len(GroupKinds))-1]
+}
+
+// GetBinaryDecision randomly returns true or false
+func GetBinaryDecision() bool {
+	return GetRandomDecision(1, 1)
+}
+
+// GetRandomDecision randomly returns true or false, but weighted to one or the other.
+// For example if success is set to 2 and failure to 1, true is twice as likely to be returned.
+func GetRandomDecision(success int, failure int) bool {
+	return rand.Intn(success+failure) > failure-1
+}
+
+// GetRandomNumber randomly returns an integer between 1 and the number specified.
+func GetRandomNumber(min int, max int) int {
+	if min == max {
+		return 1
+	} else if min > max {
+		return rand.Intn(max) + 1
+	}
+	return rand.Intn(max-min) + min + 1
+}
+
+// GetDevfile returns a structure used to represent a specific devfile in a test
+func GetDevfile(fileName string, follower DevfileFollower, validator DevfileValidator) (TestDevfile, error) {
+
+	var err error
+	testDevfile := TestDevfile{}
+	testDevfile.SchemaDevFile = schema.Devfile{}
+	testDevfile.FileName = fileName
+	testDevfile.SchemaDevFile.SchemaVersion = "2.0.0"
+	testDevfile.GroupDefaults = make(map[schema.CommandGroupKind]bool)
+	for _, kind := range GroupKinds {
+		testDevfile.GroupDefaults[kind] = false
+	}
+	testDevfile.UsedPorts = make(map[int]bool)
+	testDevfile.Validator = validator
+	testDevfile.Follower = follower
+
+	return testDevfile, err
+}
+
+// Runs a test to create and verify a devfile based on the content of the specified TestContent
+func (testDevfile *TestDevfile) RunTest(testContent TestContent, t *testing.T) {
+
+	if len(testContent.CommandTypes) > 0 {
+		numCommands := GetRandomNumber(1, maxCommands)
+		for i := 0; i < numCommands; i++ {
+			commandIndex := GetRandomNumber(1, len(testContent.CommandTypes))
+			testDevfile.AddCommand(testContent.CommandTypes[commandIndex-1])
+		}
+	}
+
+	if len(testContent.ComponentTypes) > 0 {
+		numComponents := GetRandomNumber(1, maxComponents)
+		for i := 0; i < numComponents; i++ {
+			componentIndex := GetRandomNumber(1, len(testContent.ComponentTypes))
+			testDevfile.AddComponent(testContent.ComponentTypes[componentIndex-1])
+		}
+	}
+
+	err := testDevfile.Validator.WriteAndValidate(testDevfile)
```

Comment:
I am all ears for a better way to do it.

Reply:
I looked briefly at the code but didn't easily find definitions for the various structs/interfaces used (i.e. where does `Validator` come from), but the way to do it would be to make `TestDevfile` implement the `Validator` interface:
```go
type Validator interface {
  WriteAndValidate() error
}

type TestDevfile struct {
  myField string
  mySetting bool
}

func (t *TestDevfile) WriteAndValidate() error {
  t.myField = "write"
  if !t.mySetting {
    return errors.New("Bad setting")
  }
}

// Now `TestDevfile` can be used wherever a `DevfileValidator` is needed
```

Reply:
### In common/test_utils.go, you can update:
```
// DevfileValidator interface implemented by the parser and api tests for verifying generated devfiles
type DevfileValidator interface {
	WriteAndValidate(*TestDevfile, DevfileFollower) error
}

// Runs a test to create and verify a devfile based on the content of the specified TestContent
func (testDevfile *TestDevfile) RunTest(testContent TestContent, t *testing.T, follower DevfileFollower, validator DevfileValidator) {

...

err := validator.WriteAndValidate(testDevfile, follower)
	if err != nil {
		t.Fatalf(LogErrorMessage(fmt.Sprintf("ERROR verifying devfile :  %s : %v", testContent.FileName, err)))
	}

}
```

### in api/test_utils.go, you can update:
```
// WriteAndValidate implements Saved.DevfileValidator interface.
// writes to disk and validates the specified devfile
func (devfileValidator DevfileValidator) WriteAndValidate(devfile *commonUtils.TestDevfile, follower commonUtils.DevfileFollower) error {
	err := writeDevfile(devfile)
	
	...
}


// RunTest : Runs a test to create and verify a devfile based on the content of the specified TestContent
func RunTest(testContent commonUtils.TestContent, t *testing.T) {

...

testDevfile.RunTest(testContent, t, nil, validator)
}
```

That way you dont have to embed validator and follower in TestDevfile and its much easier to read and track

Reply:
I don't understand how your versions is no much uglier than mine.

Reply:
Understand this is a test, it is meant to exercise the library api's it is not meant to be the most efficient, some places it goes out of its way to pound the api . The common code informs the follower every time something is added to the devfile, which will end up as twice for each property added to schema, for example once to add an empty command and a second to update it with attributes. This is only needed for the parser.  Once the entire devfile is created it needs to be written out and validated, which is different for the parser and the api hence the validator. This set up means the api side does not need and empty follower. This is. the pain of having common code.